tests/tests/test_provides_depends.py

```diff
@@ -0,0 +1,112 @@
+        mender_device.host_ip = enterprise_no_client.get_virtual_network_host_ip()
+
+        # Wait for ssh to be open
+        mender_device.ssh_is_opened()
+        # Check that the device has authorized with the backend.
+        device = auth_v2.get_devices(expected_devices=1)
+        device_ids = [device[0]["id"]]
+        auth_v2.accept_devices(1)
+        assert len(auth_v2.get_devices_status("accepted")) == 1
+
+        # Update client with and artifact with custom provides
+        def prepare_provides_artifact(artifact_file, artifact_id):
+            cmd = (
+                # Package MenderAPI in the artifact, just a random folder.
+                "directory-artifact-gen -o %s -n %s -t docker-client -d /tmp/test_file_update_module MenderAPI -- --provides foo:bar"
+                % (artifact_file, artifact_id)
+            )
+            logger.info("Executing: " + cmd)
+            subprocess.check_call(cmd, shell=True)
+            return artifact_file
+
+        deployment_id, _ = common_update_procedure(
+            mender_device,
+            mender_device.host_ip,
+            make_artifact=prepare_provides_artifact,
+            # We use verify_status=False, because update module updates are so
+            # quick that it sometimes races past the 'inprogress' status without
+            # the test framework having time to register it. That's not really
+            # the part we're interested in though, so just skip it.
+            verify_status=False,
+        )
+        deploy.check_expected_status("finished", deployment_id)
+
+        # Issue another update which depends on the custom provides
+        def prepare_depends_artifact(artifact_file, artifact_id):
+            cmd = (
+                # Package MenderAPI in the artifact, just a random folder.
+                "directory-artifact-gen -o %s -n %s -t docker-client -d /tmp/test_file_update_module MenderAPI -- --depends foo:bar"
+                % (artifact_file, artifact_id)
+            )
+            logger.info("Executing: " + cmd)
+            subprocess.check_call(cmd, shell=True)
+            return artifact_file
+
+        deployment_id, _ = common_update_procedure(
+            mender_device,
+            mender_device.host_ip,
+            make_artifact=prepare_depends_artifact,
+            verify_status=False,
+        )
+        deploy.check_expected_status("finished", deployment_id)
+
+        # Issue a third update with the same update as previous, this time
+        # with insufficient provides -> no artifact status
+        deployment_id, _ = common_update_procedure(
+            make_artifact=prepare_depends_artifact, verify_status=False
+        )
+
+        # Retry for at most 60 seconds checking for deployment status update
+        stat = None
+        for i in range(60):
+            time.sleep(1)
+            stat = deploy.get_statistics(deployment_id)
+            if stat.get("pending") == 0:
+                break
```

Comment:
For my own education: so there the transition will go "pending" -> "noartifact" without going through "inprogress", right?

Reply:
Correct. "inprogress" is only triggered when there is a download, which there isn't 
when "noartifact" is triggered.